sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -176,3 +179,229 @@ case class Crc32(child: Expression) extends UnaryExpression with ImplicitCastInp
     })
   }
 }
+
+/**
+ * A function that calculates hash value for a group of expressions.
+ *
+ * The hash value for an expression depends on its type:
+ *  - null:               0
```

Comment:
If we're going to change this, we should use a different value for null. Pick a large random number instead.

0 will be computed as the hash for more reasonable data (e.g. more likely an int column contains 0 than a large prime) and we can cheaply reduce some collisions.